sycl/source/detail/device_binary_image.hpp

```diff
@@ -27,9 +28,35 @@ class ByteArray {
   ConstIterator begin() const { return Ptr; }
   ConstIterator end() const { return Ptr + Size; }
 
+  template <typename... Ts> auto consume() {
+    if constexpr (sizeof...(Ts) == 1)
+      return consumeOneElem<Ts...>();
+    else
+      return std::tuple{consumeOneElem<Ts>()...};
+  }
+
+  void dropBytes(std::size_t Bytes) {
+    assert(Bytes <= Size && "Not enough bytes left!");
+    Ptr += Bytes;
+    Size -= Bytes;
+  }
+
+  template <typename T> void drop() { return dropBytes(sizeof(T)); }
+
+  bool empty() const { return Size == 0; }
+
 private:
+  template <typename T> T consumeOneElem() {
+    assert(sizeof(T) <= Size && "Out of bounds!");
+    T Val;
+    std::memcpy(&Val, Ptr, sizeof(T));
+    Ptr += sizeof(T);
+    Size -= sizeof(T);
```

Comment:
Nit, but you could call `dropBytes(sizeof(T))` here instead. Should make maintainability a little easier, in case something else comes along that needs to be mutated with every move.